zkevm-circuits/src/blake2b/circuit/src/lib.rs

```diff
@@ -0,0 +1,1855 @@
+    // of the corresponding input and output
+    pub fn compress_with_rlc(c: Value<F>, h: &[u64; 8], m: &[u64; 16], t: u128, f: bool, r: u32) -> ([u64; 8], Value<F>) {
+        let o = blake2b::compress(r, h, m, t, f);
+        let rlc = Self::compute_rlc(c, h, m, t, f, r, &o);
+        (o, rlc)
+    }
+}
+
+// An instrance of this struct describes 
+// the BLAKE2b compression function input
+#[derive(Clone, Debug)]
+pub struct CompressionInput {
+    pub r: u32, // The number of rounds
+    pub h: [u64; 8], // The initial state vector 
+    pub m: [u64; 16], // The message block vector 
+    pub t: u128, // The message byte offset
+    pub f: bool // The flag indicating the last block
+}
+
+// An instance of this struct describes the relative places of the 
+// elements of the input for a gate type of the InitialGate class 
+#[derive(Clone)]
+pub struct InitialInput<F:FieldExt> {
+    pub r: GeneralCell<F,1>, // The GeneralCell of the number of rounds 
+    pub h: [OctaChunk<F>; 8], // The array of the Chunks of the initial state vector
+    pub m: [OctaChunk<F>; 16], // The array of the Chunks of the message block vector
+    pub t: [OctaChunk<F>; 2], // The array of the Chunks of the vector (t mod 2^64, t div 2^64), where t is the message byte offset
+    pub f: BitCell<F> // The ShortCell of the flag indicating the last block
+}
+
+// An instrance of this struct describes the configuration of the circuit using R rows per round  
+#[derive(Clone)]
+pub struct CompressionConfig<F:FieldExt, const R: usize> {
+    pub rlc_table: RLCTable<F>, // The RLC table 
+    pub circuit_preparator: CircuitPreparator<F>, // The CircuitPreparator instance 
+    pub initial_gate: InitialGate<F>, // The gate type of the InitialGate class gates
+    pub round_gate: RoundGate<F>, // The gate type of the RoundGate class gates
+    pub final_gate: FinalGate<F>, // The gate type of the FinalGate class gates
+    pub initial_input: InitialInput<F>, // The relative places of the elements of the input for the gate type of the InitialGate class gates 
+    pub tail_height: usize, // The tail height 
+    pub unusable_rows: usize // The amount of the unusable rows 
+}
+
+// An instrance of this struct describes the height and input 
+// parameters of the circuit instance using R rows per round  
+#[derive(Default)]
+pub struct CompressionCircuit<F:FieldExt, const R: usize> {
+     pub k: u32, // The binary logarithm of the circuit's height 
+     pub inputs: Vec<CompressionInput>, // The BLAKE2b compression function inputs
+     pub _marker: PhantomData<F>  // The marker used to specify the circuit's native field
+}
+
+impl <F:FieldExt, const R:usize> CompressionCircuit<F,R> {
+    // Creates a CompressionCircuit for the specified binary logarithm 
+    // of the circuit's height and BLAKE2b compression function inputs 
+    pub fn new(k: u32, inputs: &[CompressionInput]) -> Self {
```

Comment:
Shouldn't `k` be a function of `inputs.len`? If not, it would be good to document in what cases we expect to have a divergent `k`. If yes, then we should probably make `k` private and simplify the API

Reply:
This circuit is designed to be a part of the supercircuit and so is going to have the height of the last one, even if such an amount of rows is not required.

Reply:
That makes sense. It could maybe be a constant instead of an argument, so the assertions would be compile-time. But that is only a small nit